src/core/thread/mlr_manager.cpp

```diff
@@ -529,6 +588,84 @@ void MlrManager::AppendToUniqueAddressList(Ip6::Address (&aAddresses)[kIPv6Addre
     return;
 }
 
+bool MlrManager::AddressListContains(const Ip6::Address *aAddressList,
+                                     uint8_t             aAddressListSize,
+                                     const Ip6::Address &aAddress)
+{
+    bool contains = false;
+
+    VerifyOrExit(aAddressListSize > 0, contains = true);
+
+    for (uint8_t i = 0; i < aAddressListSize; i++)
+    {
+        if (aAddressList[i] == aAddress)
+        {
+            ExitNow(contains = true);
+        }
+    }
+
+exit:
+    return contains;
+}
+
+void MlrManager::LogMlrResponse(otError             aResult,
+                                otError             aError,
+                                uint8_t             aStatus,
+                                const Ip6::Address *aFailedAddresses,
+                                uint8_t             aFailedAddressNum)
+{
+    OT_UNUSED_VARIABLE(aResult);
+    OT_UNUSED_VARIABLE(aError);
+    OT_UNUSED_VARIABLE(aStatus);
+    OT_UNUSED_VARIABLE(aFailedAddresses);
+    OT_UNUSED_VARIABLE(aFailedAddressNum);
+
+#if OPENTHREAD_CONFIG_LOG_BBR
+    if (aResult == OT_ERROR_NONE && aError == OT_ERROR_NONE && aStatus == ThreadStatusTlv::MlrStatus::kMlrSuccess)
+    {
+        otLogInfoBbr("Receive MLR.rsp OK", otThreadErrorToString(aResult));
+    }
+    else
+    {
+        otLogWarnBbr("Receive MLR.rsp: result=%s, error=%s, status=%d, failedAddressNum=%d",
+                     otThreadErrorToString(aResult), otThreadErrorToString(aError), aStatus, aFailedAddressNum);
+
+        for (uint8_t i = 0; i < aFailedAddressNum; i++)
+        {
+            otLogWarnBbr("MLR failed: %s", aFailedAddresses[i].ToString().AsCString());
+        }
+    }
+#endif
+}
+
+void MlrManager::CheckInvariants(void) const
+{
+#if OPENTHREAD_EXAMPLES_SIMULATION
+    uint16_t registeringNum = 0;
+
+    OT_ASSERT(!mMlrPending || mSendDelay == 0);
+
+#if OPENTHREAD_CONFIG_MLR_ENABLE
+    for (Ip6::ExternalNetifMulticastAddress &addr :
+         Get<ThreadNetif>().IterateExternalMulticastAddresses(Ip6::Address::kTypeMulticastLargerThanRealmLocal))
+    {
+        registeringNum += (addr.GetMlrState() == kMlrStateRegistering);
+    }
+#endif
+#if OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+    for (Child &child : Get<ChildTable>().Iterate(Child::kInStateValid))
+    {
+        for (const Ip6::Address &address : child.IterateIp6Addresses(Ip6::Address::kTypeMulticastLargerThanRealmLocal))
+        {
+            registeringNum += (child.GetAddressMlrState(address) == kMlrStateRegistering);
+        }
+    }
+#endif
+
+    OT_ASSERT(registeringNum == 0 || mMlrPending);
```

Comment:
```suggestion
    OT_ASSERT(registeringNum == 0 || mSendDelay != 0 || mMlrPending);
```

Reply:
This assert makes sure that if there is some MA in `Registering` state, then the device must be sending MLR.req and waiting for the response (or timeout). 
It's not directly related to `mSendDelay`. If there is a send delay but no pending MLR.req, there should be no `Registering` MAs. 
So, `registeringNum == 0 || mMlrPending` should always hold true, no need to add more conditions.

Reply:
right.